cranelift-codegen/src/isa/x86/abi.rs

```diff
@@ -585,7 +660,7 @@ fn system_v_prologue_epilogue(func: &mut ir::Function, isa: &dyn TargetIsa) -> C
     // will adjust the stack pointer to make room for the rest of the required
     // space for this frame.
     let csr_stack_size = ((csrs.iter(GPR).len() + 2) * word_size) as i32;
-    func.create_stack_slot(ir::StackSlotData {
+    let csr_slot = func.create_stack_slot(ir::StackSlotData {
         kind: ir::StackSlotKind::IncomingArg,
         size: csr_stack_size as u32,
         offset: Some(-csr_stack_size),
```

Comment:
I was just looking at this for the same reason, I think you're on the right track. These offsets are up from the bottom of the stack slot so zero is on the opposite side from where `push` starts placing registers. My expectation was the callee-save region would look like
```
size      | rbp
size - 8  | rbx
size - 16 | rsi
... other callee-save gprs
0x20      | xmm8
0x10      | xmm7
0x00      | xmm6
```
which is close to what the actual layout looks like, but the base isn't what I'd expected.

Reply:
update: it's not too late, but `layout_stack` requires slots of type `IncomingArg` to have a defined offset. That offset is used for the base of the stack layout so a `None` ends up being a panic.

I ended up making FPR save/restore go through a second stack slot of kind `SpillSlot` - it seems like either this or `ExplicitSlot` might make sense on account of stack store/load is a lot like a spill (if they weren't used, they wouldn't have to be "spilled" to the stack). I'm not entirely sure what `ExplicitSlot` ought to be used for, but it also looks applicable?

Reply:
I'm not totally sure, but I *think* that `SpillSlot`s are intended to be reused/de-duped across spilled values with distinct live ranges. This should be possible to do in theory for calling convention related stack slots (same as with struct return) but I'm not sure that it works as of today nor that it wouldn't be buggy.

I think (again not totally sure, as they're mostly used in tests and pretty much all slot kinds other than incoming args are treated the same way) that explicit slots are intended as catch all slots used for random purposes.

It seems that other csr slots are all `IncomingArg` tho, so maybe its best to keep that slot kind and figure out what the explicit offset needs to be, rather than rely on `layout_stack`.

Reply:
I think we'd need to make some changes to `stack_layout` and `StackSlot` for "correct offset in `StackSlotKind::IncomingArg`" to be a question we can really answer. `stack_addr` and friends assume the prologue has already done its stack setup, and, seemingly, that `rsp` won't move through the body of the function. So for the address of an `IncomingArg` slot to be used _in_ the prologue, we'd need to defer actual resolution of stack offsets even after the instructions are added.

The other option is tracking the offset ourselves in the prologue, which I expect would look like tracking the amount we've moved SP with pushes/explicit adjustment, then acknowledging that the address from `stack_addr(IncomingArgSlot)` will be wrong (it'd be off by however much we've pushed) and correcting for that after the fact. Because floating point save/restore is past any GPR pushes, I think the offset will always be valid, just smaller than it "should" be.

I'd be more suspicious that I'm misunderstanding something, but I don't think we have other instances of wanting to do stack loads/stores in prologues/epilogues, so maybe this is actually a rough spot? :(